### ISSUE TITLE
Mysterious parallelism

### ISSUE THREAD

**Issue by gokceneraslan** [LTLA/SingleR#62]
Hi,

Even if I run SingleR with `BPPARAM=BiocParallel::SerialParam()`, I see multiple processes working. Since I'm running SingleR in docker with limited memory, this usually leads to a memory error and consequently the following error:

```
Error in result[[njob]] <- value : 
  attempt to select less than one element in OneIndex
Calls: <Anonymous> ... .local -> bplapply -> bplapply -> bploop -> bploop.lapply

In parallel::mccollect(wait = FALSE, timeout = 1) :
1 parallel job did not deliver a result
```

I was wondering if you know whether there is another source of parallelism in the code :/

**Comment by LTLA** [comment]
There should not be.

**Comment by LTLA** [comment]
Absent a bug in **BiocParallel**, I'm confident that **SingleR** is not the cause of your error. You don't provide any context so I can only sugges that you check your other function calls, possibly by setting `register(SerialParam())` at the top to avoid inadvertent parallelization of **DelayedArray** functions.

**Comment by gokceneraslan** [comment]
Thanks, I'll try that.

**Comment by gokceneraslan** [comment]
Even after calling "BiocParallel::register(BiocParallel::SerialParam())" and calling SingleR function with `BPPARAM=BiocParallel::SerialParam()`, it still creates 6 processes towards the end of annotation process and crashes if the matrix is really big. I'm super puzzled. I'm not sure if it's relevant but I'm calling SingleR from python through rpy2.

**Comment by gokceneraslan** [comment]
Here is the latest error message I have:

```
R[write to console]: snapshotDate(): 2019-10-22

R[write to console]: see ?SingleR and browseVignettes('SingleR') for documentation

R[write to console]: loading from cache

R[write to console]: see ?SingleR and browseVignettes('SingleR') for documentation

R[write to console]: loading from cache

R[write to console]: snapshotDate(): 2019-10-22

R[write to console]: see ?SingleR and browseVignettes('SingleR') for documentation

R[write to console]: loading from cache

R[write to console]: see ?SingleR and browseVignettes('SingleR') for documentation

R[write to console]: loading from cache

R[write to console]: Error in result[[njob]] <- value :
  attempt to select less than one element in OneIndex

R[write to console]: In addition:
R[write to console]: Warning message:

R[write to console]: In parallel::mccollect(wait = FALSE, timeout = 1) :
R[write to console]:

R[write to console]:  1 parallel job did not deliver a result
```

**Comment by LTLA** [comment]
Does this happen if you run R normally? The multiple `see ?SingleR and browseVignettes('SingleR') for documentation` calls are sort of weird, unless you're loading multiple references.

**Comment by gokceneraslan** [comment]
I'm loading multiple references indeed. I have never tried running it in R normally. Here is the gist of my horrible code, if it helps: https://gist.github.com/gokceneraslan/f953b697ea09b9a8ff890a8707775498

**Comment by LTLA** [comment]
One thing to try is also:

```r
setAutoBPPARAM(SerialParam())
```

to tell **DelayedArray** to only use a single core. I don't recall whether we do any DA operations, but it's worth a shot.

**Comment by LTLA** [comment]
Any news? Is it still failing with the DA settings?

**Comment by LTLA** [comment]
I'm about 80% sure that this is fixed by 9ba36b9e67d9e3d61715b7e0a4eabf219d3d418d, so I'll close for now.
